torch/distributed/_shard/sharding_plan/api.py

```diff
@@ -0,0 +1,63 @@
+import abc
+import torch.nn as nn
+
+from dataclasses import dataclass
+from typing import Dict, List, Optional
+
+from torch.distributed._shard.sharding_spec import ShardingSpec
+
+@dataclass
+class ShardingPlan(object):
+    """
+    Representation of a sharding plan, describes how to shard a module
+    across hosts.
```

Comment:
Maybe extend these docs a bit elaborating on `plan`, `output_plan` and `collect_local_shards` and give an example of how they might be used together.